chain/client/src/view_client.rs

```diff
@@ -229,50 +219,31 @@ impl ViewClientActor {
             QueryRequest::ViewCode { account_id, .. } => account_id,
         };
         let shard_id = self.runtime_adapter.account_id_to_shard_id(account_id);
+        let head = self.chain.head()?;
 
-        // If we have state for the shard that we query return query result directly.
-        // Otherwise route query to peers.
-        match self.chain.get_chunk_extra(header.hash(), shard_id) {
-            Ok(chunk_extra) => {
-                let state_root = chunk_extra.state_root;
-                self.runtime_adapter
-                    .query(
-                        shard_id,
-                        &state_root,
-                        header.height(),
-                        header.raw_timestamp(),
-                        header.prev_hash(),
-                        header.hash(),
-                        header.epoch_id(),
-                        &msg.request,
-                    )
-                    .map(Some)
-                    .map_err(|e| e.to_string())
-            }
-            Err(e) => {
-                match e.kind() {
-                    ErrorKind::DBNotFoundErr(_) => {}
-                    _ => {
-                        warn!(target: "client", "Getting chunk extra failed: {}", e.to_string());
-                    }
-                }
-                // route request
-                let mut request_manager = self.request_manager.write().expect(POISONED_LOCK_ERR);
-                if Self::need_request(msg.query_id.clone(), &mut request_manager.query_requests) {
-                    let validator = self
-                        .chain
-                        .find_validator_for_forwarding(shard_id)
-                        .map_err(|e| e.to_string())?;
-                    self.network_adapter.do_send(NetworkRequests::Query {
-                        query_id: msg.query_id,
-                        account_id: validator,
-                        block_reference: msg.block_reference,
-                        request: msg.request,
-                    });
-                }
-
-                Ok(None)
-            }
+        if self.runtime_adapter.cares_about_shard(
+            self.validator_account_id.as_ref(),
+            &head.last_block_hash,
+            shard_id,
+            true,
+        ) {
+            let chunk_extra = self.chain.get_chunk_extra(header.hash(), shard_id)?;
+            let state_root = chunk_extra.state_root;
+            Ok(self
+                .runtime_adapter
+                .query(
+                    shard_id,
+                    &state_root,
+                    header.height(),
+                    header.raw_timestamp(),
+                    header.prev_hash(),
+                    header.hash(),
+                    header.epoch_id(),
+                    &msg.request,
+                )
+                .map_err(RuntimeQueryError::from)?)
```

Comment:
I wonder if it is the only place where we need this `RuntimeQueryError`. If so, it might be easier to move the `match` that is currently in the `from` implementation right here inline and avoid the `RuntimeQueryError`. This will reduce the indirection, but will bloat this implementation :disappointed: 

WDYT?